src/sessions.ts

```diff
@@ -764,20 +737,37 @@ function endTransaction(
       }
     }
 
-    callback(error);
+    if (error != null) {
+      throw error;
+    }
   }
 
   if (session.transaction.recoveryToken) {
     command.recoveryToken = session.transaction.recoveryToken;
   }
 
-  const handleFirstCommandAttempt = (error?: Error) => {
+  const partiallyHandleFirstCommandAttempt = async () => {
     if (command.abortTransaction) {
       // always unpin on abort regardless of command outcome
       session.unpin();
     }
+  };
 
-    if (error instanceof MongoError && isRetryableWriteError(error)) {
+  try {
+    // send the command
+    await executeOperation(
+      session.client,
+      new RunAdminCommandOperation(command, {
+        session,
+        readPreference: ReadPreference.primary,
+        bypassPinningCheck: true
+      })
+    );
+    await partiallyHandleFirstCommandAttempt();
+    commandHandler();
```

Comment:
Can we remove `commandHandler` in favor of inlining the logic? it is still callback-like

Reply:
The logic for command handler is called 4 separate times, which is why i kept it as a function. To make it less callback like, I can change its name to not include 'handler.' Lmk what you think!

Reply:
It's called 4 times  but its logic is conditional, we don't need the exact same code to run in all of these places when a portion of it will never be entered and if we need to re-run some of the same code we can employ a loop. Both tatics will reduce the amount of code executed, and remove a function created on each invocation of `endTransaction`. I think it is worth breaking it out here for that benefit.

Reply:
Bookkeeping from Slack DM discussion: 
- Non-error case is moved to inline
- Error case function is no longer nested and renamed from `commandHandler` to `handleEndTransactionError`